### ISSUE TITLE
Take current DI to the next level

### ISSUE THREAD

**Issue by runt9** [runt9/FusionOfSouls#33]
Currently our Dependency Injection is functional, but must be manually maintained and defined. It would be nice to utilize Dependency Injection in a more automated way.

In addition, the paradigm throughout the game should be "Everything that **can** be Dependency Injected **should** be dependency injected"

**Comment by runt9** [comment]
The main source of pain was having to update the bind call each time an injected field was added or removed, and this simple block of code solves that:

```
inline fun <reified Type : Any> Context.autoBindSingleton() = bindSingleton(autoInject<Type>())

inline fun <reified T> autoInject(): T {
    val constructor = T::class.constructors.first()
    val args = constructor.parameters.map { param -> injector.getProvider(param.type.jvmErasure.java)() }.toTypedArray()
    return constructor.call(*args)
}
```

Auto-scanning and managing the dependency graph is probably out-of-scope for now as it's not a huge pain point to just make sure things are registered with the injector and in the right order since the auto-inject takes care of the rest

**Comment by runt9** [comment]
Closed by multiple commits, confirmed closed by 94d2adc74adf28376a4a23b70fb69bb723245fbc
